tools/dbslower.py

```diff
@@ -39,21 +46,87 @@
     help="the database engine to use")
 parser.add_argument("-p", "--pid", type=int, nargs='*',
     dest="pids", metavar="PID", help="the pid(s) to trace")
+parser.add_argument("-b", "--binary", type=str,
+    dest="path", metavar="PATH", help="path to binary")
 parser.add_argument("-m", "--threshold", type=int, default=1,
     help="trace queries slower than this threshold (ms)")
 args = parser.parse_args()
 
-if not args.pids or len(args.pids) == 0:
-    if args.db == "mysql":
-        args.pids = map(int, subprocess.check_output(
-                                        "pidof mysqld".split()).split())
-    elif args.db == "postgres":
-        args.pids = map(int, subprocess.check_output(
-                                        "pidof postgres".split()).split())
-
 threshold_ns = args.threshold * 1000000
 
-program = """
+program_uprobe = """
+#include <uapi/linux/ptrace.h>
+
+struct temp_t {
+    u64 timestamp;
+    char query[256];
+};
+
+struct data_t {
+    u64 pid;
+    u64 timestamp;
+    u64 duration;
+    char query[256];
+};
+
+BPF_HASH(temp, u64, struct temp_t);
+BPF_PERF_OUTPUT(events);
+
+int mysql56_dispatch_start(struct pt_regs *ctx) {
+    u64 pid = bpf_get_current_pid_tgid();
+    u64 command  = (u64) PT_REGS_PARM1(ctx);
+
+    if (command == 3) {
+        struct temp_t tmp = {};
+        tmp.timestamp = bpf_ktime_get_ns();
+        bpf_probe_read(&tmp.query, sizeof(tmp.query), (void*) PT_REGS_PARM3(ctx));
+
+        temp.update(&pid, &tmp);
+    }
+    return 0;
+};
+
+int mysql57_dispatch_start(struct pt_regs *ctx) {
+    u64 pid = bpf_get_current_pid_tgid();
+    u64 command  = (u64) PT_REGS_PARM3(ctx);
+
+    if (command == 3) {
+        struct temp_t tmp = {};
+        tmp.timestamp = bpf_ktime_get_ns();
+
+        void* st = (void*) PT_REGS_PARM2(ctx);
+        char* query;
+        bpf_probe_read(&query, sizeof(query), st);
+        bpf_probe_read(&tmp.query, sizeof(tmp.query), query);
+
+        temp.update(&pid, &tmp);
+    }
+    return 0;
+};
+
+int dispatch_end(struct pt_regs *ctx) {
+    struct temp_t *tempp;
+    u64 pid = bpf_get_current_pid_tgid();
+    tempp = temp.lookup(&pid);
+    if (!tempp)
+        return 0;
+
+    u64 delta = bpf_ktime_get_ns() - tempp->timestamp;
+    if (delta >= """ + str(threshold_ns) + """) {
+        struct data_t data = {};
+        data.pid = pid >> 32;   // only process id
+        data.timestamp = tempp->timestamp;
+        data.duration = delta;
+        bpf_probe_read(&data.query, sizeof(data.query), tempp->query);
+        events.perf_submit(ctx, &data, sizeof(data));
+    }
+    temp.delete(&pid);
+    return 0;
+};
+
+"""
+
+program_udst = """
```

Comment:
There's quite a bit of code duplication, especially here around the result reporting. Perhaps instead of having two completely separate programs, you could use conditional compilation or text replacement to have as much shared code as possible? When the scripts grow, any duplication becomes a maintenance burden.